SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp

```diff
@@ -792,10 +861,106 @@ void USpatialReceiver::ApplyComponentData(Worker_EntityId EntityId, Worker_Compo
+		// If we can't find this subobject, it's a dynamically attached object. Create it now.
+		TargetObject = NewObject<UObject>(Channel->GetActor(), ClassInfoManager->GetClassByComponentId(Data.component_id));
+
+		Channel->GetActor()->OnSubobjectCreatedFromReplication(TargetObject.Get());
+
+		PackageMap->ResolveSubobject(TargetObject.Get(), FUnrealObjectRef(EntityId, Offset));
+
+		Channel->CreateSubObjects.Add(TargetObject.Get());
+	}
+
+	ApplyComponentData(TargetObject.Get(), Channel, Data);
+}
+
+void USpatialReceiver::HandleDynamicAddComponent(const Worker_AddComponentOp& Op)
+{
+	uint32 Offset = 0;
+	bool bFoundOffset = ClassInfoManager->GetOffsetByComponentId(Op.data.component_id, Offset);
+	if (!bFoundOffset)
+	{
+		UE_LOG(LogSpatialReceiver, Warning, TEXT("EntityId %lld, ComponentId %d - Could not find offset for component id "
+			"when receiving dynamic AddComponent."), Op.entity_id, Op.data.component_id);
+		return;
+	}
+
+	// Object already exists, we can apply data directly. 
+	if (UObject* Object = PackageMap->GetObjectFromUnrealObjectRef(FUnrealObjectRef(Op.entity_id, Offset)).Get())
+	{
+		ApplyComponentData(Object, NetDriver->GetActorChannelByEntityId(Op.entity_id), Op.data);
+		return;
+	}
+
+	// Otherwise this is a dynamically attached component. We need to make sure we have all related components before creation.
+	PendingDynamicSubobjectComponents.Add(MakeTuple(static_cast<Worker_EntityId_Key>(Op.entity_id), Op.data.component_id),
+		PendingAddComponentWrapper(Op.entity_id, Op.data.component_id, MakeUnique<DynamicComponent>(Op.data)));
+
+	const FClassInfo& Info = ClassInfoManager->GetClassInfoByComponentId(Op.data.component_id);
+
+	bool bReadyToCreate = true;
+	ForAllSchemaComponentTypes([&](ESchemaComponentType Type)
+	{
+		Worker_ComponentId ComponentId = Info.SchemaComponents[Type];
+
+		if (ComponentId == SpatialConstants::INVALID_COMPONENT_ID)
+		{
+			return;
+		}
+
+		if (!PendingDynamicSubobjectComponents.Contains(MakeTuple(static_cast<Worker_EntityId_Key>(Op.entity_id), ComponentId)))
+		{
+			bReadyToCreate = false;
+		}
+	});
+
+	if (bReadyToCreate)
+	{
+		AttachDynamicSubobject(Op.entity_id, Info);
+	}
+}
+
+void USpatialReceiver::AttachDynamicSubobject(Worker_EntityId EntityId, const FClassInfo& Info)
+{
+	AActor* Actor = Cast<AActor>(PackageMap->GetObjectFromEntityId(EntityId).Get());
+
+	if (Actor == nullptr)
+	{
+		UE_LOG(LogSpatialReceiver, Warning, TEXT("Tried to dynamically attach subobject of type %s to entity %lld but couldn't find Actor!"), *Info.Class->GetName(), EntityId);
 		return;
 	}
 
+	UObject* Subobject = NewObject<UObject>(Actor, Info.Class.Get());
+
+	Actor->OnSubobjectCreatedFromReplication(Subobject);
+
+	PackageMap->ResolveSubobject(Subobject, FUnrealObjectRef(EntityId, Info.SchemaComponents[SCHEMA_Data]));
+
+	USpatialActorChannel* Channel = NetDriver->GetActorChannelByEntityId(EntityId);
```

Comment:
If you've received a AttachDynamicSubobject late then it's possible this actor could be torn off and not have a channel.

Reply:
Changed.